src/main/java/net/dv8tion/jda/api/components/utils/ComponentPathIterator.java

```diff
@@ -0,0 +1,165 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.api.components.utils;
+
+import net.dv8tion.jda.api.components.Component;
+import net.dv8tion.jda.api.components.actionrow.ActionRow;
+import net.dv8tion.jda.api.components.container.Container;
+import net.dv8tion.jda.api.components.section.Section;
+import org.apache.commons.collections4.iterators.SingletonIterator;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Utility class to recursively iterate on a list of components while keeping track their path.
+ */
+public class ComponentPathIterator implements Iterator<ComponentPathIterator.ComponentWithPath>
+{
+    private final Stack<Iterator<ComponentWithPath>> stack = new Stack<>();
+
+    protected ComponentPathIterator(String rootName, Collection<? extends Component> components) {
+        stack.push(new CollectionAttributeIterator(rootName, "components", components));
+    }
+
+    /**
+     * Creates a {@link ComponentPathIterator} to recursively iterate on the provided components.
+     *
+     * @param  components
+     *         The components to iterate on
+     *
+     * @return A new {@link ComponentPathIterator}
+     */
+    @Nonnull
+    public static ComponentPathIterator create(@Nonnull String rootName, @Nonnull Collection<? extends Component> components) {
+        return new ComponentPathIterator(rootName, components);
+    }
+
+    /**
+     * Creates a {@link Stream} of {@link ComponentWithPath} which recursively iterates on the provided components.
+     *
+     * @param  components
+     *         The components to iterate on
+     *
+     * @return A new, ordered {@link Stream} of {@link ComponentWithPath}
+     */
+    @Nonnull
+    public static Stream<ComponentWithPath> createStream(@Nonnull String rootName, @Nonnull Collection<? extends Component> components) {
+        Spliterator<ComponentWithPath> spliterator = Spliterators.spliteratorUnknownSize(create(rootName, components), Spliterator.ORDERED);
+        return StreamSupport.stream(spliterator, false);
+    }
+
+    @Override
+    public boolean hasNext() {
+        ensureNestedIteratorHasNext();
+        return !stack.isEmpty();
+    }
+
+    @Nonnull
+    @Override
+    public ComponentWithPath next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        Iterator<ComponentWithPath> iterator = stack.peek();
+        ComponentWithPath componentWithPath = iterator.next();
+        Component component = componentWithPath.component;
+
+        if (component instanceof Container) {
```

Comment:
Let's add snapshot tests for both of these iterators, so this flattening logic is not forgotten when new components are added.